crypto/rsa/rsa_sign.c

```diff
@@ -112,106 +116,91 @@ int int_rsa_verify(int dtype, const unsigned char *m,
     }
-    if ((dtype == NID_md5_sha1) && (m_len != SSL_SIG_LENGTH)) {
-        RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_INVALID_MESSAGE_LENGTH);
-        goto err;
-    }
-    i = RSA_public_decrypt((int)siglen, sigbuf, s, rsa, RSA_PKCS1_PADDING);
 
+    i = RSA_public_decrypt((int)siglen, sigbuf, s, rsa, RSA_PKCS1_PADDING);
     if (i <= 0)
         goto err;
-    /*
-     * Oddball MDC2 case: signature can be OCTET STRING. check for correct
-     * tag and length octets.
-     */
-    if (dtype == NID_mdc2 && i == 18 && s[0] == 0x04 && s[1] == 0x10) {
-        if (rm) {
-            memcpy(rm, s + 2, 16);
-            *prm_len = 16;
-            ret = 1;
-        } else if (memcmp(m, s + 2, 16)) {
-            RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_BAD_SIGNATURE);
-        } else {
-            ret = 1;
-        }
-    } else if (dtype == NID_md5_sha1) {
-        /* Special case: SSL signature */
-        if ((i != SSL_SIG_LENGTH) || memcmp(s, m, SSL_SIG_LENGTH))
-            RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_BAD_SIGNATURE);
-        else
-            ret = 1;
-    } else {
-        const unsigned char *p = s;
-        sig = d2i_X509_SIG(NULL, &p, (long)i);
 
-        if (sig == NULL)
+    if (dtype == NID_md5_sha1) {
+	/* NID_md5_sha1 corresponds to the MD5/SHA1 combination in TLS 1.1 and
+	 * earlier. It has no DigestInfo wrapper but otherwise is
+	 * RSASSA-PKCS1-v1_5. */
+        if (i != SSL_SIG_LENGTH) {
+            RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_BAD_SIGNATURE);
             goto err;
+        }
 
-        /* Excess data can be used to create forgeries */
-        if (p != s + i || !rsa_check_digestinfo(sig, s, i)) {
+        if (rm != NULL) {
+            memcpy(rm, s, SSL_SIG_LENGTH);
+            *prm_len = SSL_SIG_LENGTH;
+        } else if (memcmp(s, m, SSL_SIG_LENGTH) != 0) {
             RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_BAD_SIGNATURE);
             goto err;
         }
-
+    } else if (dtype == NID_mdc2 && i == 18 && s[0] == 0x04 && s[1] == 0x10) {
         /*
-         * Parameters to the signature algorithm can also be used to create
-         * forgeries
+         * Oddball MDC2 case: signature can be OCTET STRING. check for correct
+         * tag and length octets.
          */
-        if (sig->algor->parameter
-            && ASN1_TYPE_get(sig->algor->parameter) != V_ASN1_NULL) {
+        if (rm != NULL) {
+            memcpy(rm, s + 2, 16);
+            *prm_len = 16;
+        } else if (memcmp(m, s + 2, 16) != 0) {
             RSAerr(RSA_F_INT_RSA_VERIFY, RSA_R_BAD_SIGNATURE);
             goto err;
         }
+    } else {
+        /*
+         * In verifyrecover mode, construct the encoded form using the end of
+         * the structure. If the DigestInfo prefix is invalid, the signature
+         * check will cover it.
+         */
+        if (rm != NULL) {
+            const EVP_MD *md = EVP_get_digestbynid(dtype);
+            m_len = EVP_MD_size(md);
+            if (md != NULL && m_len > (size_t)i) {
```

Comment:
EVP_MD_size() can actually handle NULL, but it seems to be a bad style to first call a function with NULL and then later check that it the variable was NULL and give an error.